src/Core/src/Signer/SignerV4.php

```diff
@@ -56,87 +57,202 @@ public function __construct(string $scopeName, string $region)
+                self::ALGORITHM_REQUEST,
+                $credentials->getAccessKeyId(),
+                implode('/', $credentialScope),
+                implode(';', \array_keys($canonicalHeaders)),
+                $signature
+            ));
+        }
+    }
+
+    private function removePresign(Request $request): void
+    {
+        $request->removeQueryAttribute('X-Amz-Algorithm');
+        $request->removeQueryAttribute('X-Amz-Signature');
+        $request->removeQueryAttribute('X-Amz-Security-Token');
+        $request->removeQueryAttribute('X-Amz-Date');
+        $request->removeQueryAttribute('X-Amz-Expires');
+        $request->removeQueryAttribute('X-Amz-Credential');
+        $request->removeQueryAttribute('X-Amz-SignedHeaders');
+    }
+
+    private function sanitizeHostForHeader(Request $request): void
+    {
         if (false === $parsedUrl = parse_url($request->getEndpoint())) {
             throw new InvalidArgument(sprintf('The endpoint "%s" is invalid.', $request->getEndpoint()));
         }
+
+        $host = $parsedUrl['host'];
+        if (isset($parsedUrl['port'])) {
+            $host .= ':' . $parsedUrl['port'];
+        }
+
+        $request->setHeader('host', $host);
+    }
+
+    private function assignAmzQueryValues(Request $request, Credentials $credentials, bool $isPresign): void
+    {
+        if ($isPresign) {
+            $request->setQueryAttribute('X-Amz-Algorithm', self::ALGORITHM_REQUEST);
+            if (null !== $sessionToken = $credentials->getSessionToken()) {
+                $request->setQueryAttribute('X-Amz-Security-Token', $sessionToken);
+            }
+
+            return;
+        }
+
         if (null !== $sessionToken = $credentials->getSessionToken()) {
             $request->setHeader('x-amz-security-token', $sessionToken);
         }
+    }
 
-        $request->setHeader('host', $parsedUrl['host'] . (isset($parsedUrl['port']) ? ':' . $parsedUrl['port'] : ''));
-        $request->setHeader('x-amz-date', $amzDate = gmdate('Ymd\THis\Z'));
-        $credentialScope = [substr($amzDate, 0, 8), $this->region, $this->scopeName, 'aws4_request'];
+    private function buildTime(Request $request, \DateTimeInterface $now, \DateTimeInterface $expires, bool $isPresign): void
+    {
+        if ($isPresign) {
+            $duration = $expires->getTimestamp() - $now->getTimestamp();
+            if ($duration > 604800) {
+                throw new InvalidArgument('The expiration date of presigned URL must be less than one week');
+            }
+            if ($duration < 0) {
+                throw new InvalidArgument('The expiration date of presigned URL must be in the future');
+            }
 
-        $signingKey = 'AWS4' . $credentials->getSecretKey();
-        foreach ($credentialScope as $scopePart) {
-            $signingKey = hash_hmac('sha256', $scopePart, $signingKey, true);
+            $request->setQueryAttribute('X-Amz-Date', gmdate('Ymd\THis\Z', $now->getTimestamp()));
```

Comment:
Why not using `$now-format()` here ?

Reply:
Fixed. 
Added a `->setTimezone('GMT')` in `sign`/`presign` methods